### ISSUE TITLE
Simple minded function approximation

### ISSUE THREAD

**Issue by Polyacene** [tiny-dnn/tiny-dnn#649]
Dear All,
in trying to learn a little bit about neural networks I came across tiny-dnn,
which appears to be nicely suited for someone familiar with C++.
As an initial test I tried to construct a network that approximates a simples function,
5*x exp( -5*x^2) in  the range (-1:1). However I failed miserably, the predicted values 
do not reflect the desired function.
May someone give me (== an absolute neural network beginner), a hint?, I'm a little bit clueless.
The code is below.

Thanks in advance,
best regards,
Peter

```C++
// g++ -O3 -march=native  -pthread -std=c++11 -I ../tiny-dnn/ Train.C -o Train

#include <stdlib.h>
#include <iostream>
#include <vector>

#include <tiny_dnn/tiny_dnn.h>

typedef float TpFloat;

using namespace tiny_dnn;
using namespace tiny_dnn::activation;

TpFloat UnknownFunction( const TpFloat x )
{
	return 5*x*exp( -5*x*x ) ;
}

int main(int argc, char** argv) {
  
	const unsigned int NumHidden = (argc > 1 ) ? atoi( argv[1] ) : 15;
	const unsigned int NumTrain  = (argc > 2 ) ? atoi( argv[2] ) : 10000;

	int n_batch = (argc > 3 ) ? atoi( argv[3] ) : 25;
	int n_epoch = (argc > 4 ) ? atoi( argv[4] ) : 30;
	
	int NumEvaluate = 25;
	
	network<sequential> nn = make_mlp<tan_h>({ 1, NumHidden, 1 });

	adagrad optimizer;
	//adadelta optimizer;

	std::vector<vec_t> input_data;
	std::vector<vec_t> desired_out;
	
	{ 	/// generate input
		vec_t  r(1), v(1);
		
		for( int n=0; n<NumTrain; ++n )
		{
			r[0] = 1.999 *(drand48() - 0.5);
			v[0] = UnknownFunction( r[0] );
			input_data.push_back( r );
			desired_out.push_back( v );
		}
	}
	std::cout << "# " << NumTrain << " data points generated for " << NumHidden << " neurons." << std::endl;
	
	std::cout << "start training witch batch: " << n_batch << " epoch: " << n_epoch << std::endl;
	nn.train<cross_entropy>( optimizer, input_data, desired_out, n_batch, n_epoch);

	std::cout << "# Evaluate:" << std::endl;
	
	{ /// generate input
		vec_t  p(1);
		
		for( int n=0; n<NumEvaluate; ++n )
		{
			p = nn.predict( input_data[n] );
			std::cout << input_data[n][0] << '\t' << p[0] << '\t' << desired_out[n][0] << std::endl;
		}
	}
	
	std::cout << "\n# Test:" << std::endl;
	{ /// generate input
		vec_t  r(1), v(1),p(1);
		
		for( int n=0; n<NumEvaluate; ++n )
		{
			r[0] = 2*drand48() -1.0;
			v[0] = UnknownFunction( r[0] );
			p = nn.predict( r );
			
			std::cout << r[0] << '\t' << p[0] << '\t' << v[0] << std::endl;
		}
	}

	return 0;
}
```

**Comment by pliptor** [comment]
@Polyacene  Before adding more layers, try these two following changes and leave everything else untouched. You should gain some insights:

1. Replace `tan_h` in your make_mpl by `relu`
2. Replace `cross_entropy` in your train function by `mse`

**Comment by Polyacene** [comment]
@Randl: Adding Layers didn't help.

@Pliptor: Using relu & mse leads to results which are not completely off.
  I also changed the domain to (0,1). Looks like one shouldn't come too close to 0 and 1.
Restricting the domain further to (0.1, 0.9) it looks not too bad for some values, but sometimes they are pretty off. Which makes we wonder, whether I messed up something fundamentally. After all,
my examples is a pretty smooth and boring function.

**Comment by pliptor** [comment]
@Polyacene Did you notice the positive values are better but the negative values are estimated as zero?
If yes, that was part 1 of the exercise.

**Comment by Polyacene** [comment]
@pliptor : Yes, that's why I changed the domain. BTW, tan_h works also better when restricting to
  x values in (-0.5:0.5).

**Comment by pliptor** [comment]
@Polyacene Good. So if your goal was to learn what is going on, I hope you now understand the choice of "activations" affect the range of input/output representation. Also the type of distortions they suffer.

**Comment by Polyacene** [comment]
@pliptor : Well, I understand the concept of input values in general. However, I thought that
 the network should decompose my input function into linear combinations of the activation
functions. And even if that doesn't work too well I didn't expect deviations which are typically > 50%,
and sometimes even significantly larger.
So I wondered, whether I made some fundamental error.

**Comment by pliptor** [comment]
The biggest fundamental problem, I think, was the choice of cross_entropy for the loss function. I believe it is more often used in classification tasks using multiple output neurons For regression, mse might be a better loss function. Also @Randl is right, after the above changes, adding another layer should further improve precision. But I thought it was helpful to squeeze as much as possible with fewer neurons and layers first.

**Comment by pliptor** [comment]
@Polyacene I thought your idea of experimenting how well a neural network interpolates a non-linear curve was very interesting so I tried a few things and got something to share. The attached code has the following features.

```
1. Single hidden layer, 1 input and 1 output neuron like yours, but using only ReLU and identity activation  functions at the output for maximal network structure simplicity (a ReLU function is just a ramp function).
2. 15 hidden neurons like yours.
3. Tested the input ranges [-2, 2] and it handles okay everywhere.
4. Tested larger output ranges (up to [0, 8] with an optional quadratic function) and it handles okay everywhere as well.
```
I hope you find it useful.

[interpolate.cpp.txt](https://github.com/tiny-dnn/tiny-dnn/files/908971/interpolate.cpp.txt)

**Comment by edgarriba** [comment]
Please check this Keras example, mse loss is also an option for regression problems
http://machinelearningmastery.com/regression-tutorial-keras-deep-learning-library-python/

@Polyacene @pliptor could you send a PR as a regression example? The cpp with a small readme explanation, plots can be very useful for new users.

**Comment by Polyacene** [comment]
@pliptor Thanks a lot. Your example really makes a difference and provides a great starting point.
Is there a specific reason for the linear activation in the final layer?
I've attached a plot, [Res.pdf](https://github.com/tiny-dnn/tiny-dnn/files/909628/Res.pdf), of the results, and one where I replaced relu by tanh_h. As expected, tanh_h 
provides a smoother prediction. In the attached plot N stands for NumHidden and N_T for
the number of training samples. I didn't change n_batch and n_epoch.
I'll take my time to perform a more systematic investigation. 

@edgarriba I would provide a PR with examples. However I'll need some time for that. Besides
being still an old-fashioned svn person having to learn git first, I feel a little bit uncomfortable
providing a PR without some basic understanding on what I'm actually doing. At least I want 
to perform some systematic studies first.

**Comment by pliptor** [comment]
@Polyacene I have no theory backing for the linear activation in the final layer. I just thought it would help the network with unbounded input/output. In principle, I think there is nothing in the code itself or at least what I understand what the algorithms do that force the input or output to be bounded unless you use activations that do so. If you Google search you'll find typical recommendations to normalize and bound the input though (and that's what I would normally do). Sometimes I insert linear activations to tame the network a bit though.

You are right, intuitively tan_h provides a better fit in these examples and the graphs also show that.  I just like ReLUs and try to use them whenever I can :) You'll find about them soon.

**Comment by Polyacene** [comment]
@pliptor Here's my first scaling, 
[Reg.pdf](https://github.com/tiny-dnn/tiny-dnn/files/910795/Reg.pdf)
performed in simple way by just looping over nn.train(...).
Looks like it's better to increase the training ensemble and <mse> seems
to perform better than <absolute>. Have to stop for today.
All data set are for an tanh activation.

**Comment by pliptor** [comment]
@Polyacene Thanks for sharing the graph. It's a nice graph. It shows clearly the log-log dependency of accuracy versus epoch for this example. I believe the noise floor comes from the quantization of the training ensemble and their levels seems to match the resolution difference.

**Comment by pliptor** [comment]
@Polyacene Did you do any other interesting studies?

**Comment by Polyacene** [comment]
@pliptor I've been playing a little bit with higher dimensional functions, which turned out to "kind of work". At least I'm now much more comfortable in using tiny-dnn. I found where to look for optimizers in the source code and learned how to use the call-back functions in fit(...).

**Comment by Thanh-Binh** [comment]
Could you share your code for function fit?

**Comment by Polyacene** [comment]
@Thanh-Binh I've attached, [interpolate.C.txt](https://github.com/tiny-dnn/tiny-dnn/files/954218/interpolate.C.txt), a sample code for fitting, which incorporates suggestions from
pliptor. The main problem in my original version was the choice of the cost function. 
The extension to higher dimensional functions is straightforward. Just increase the size of the input vectors.

**Comment by Thanh-Binh** [comment]
@polyacene thank for sharing your code. Do you have any comparison result: fitting accuracy, how quickly it learns?

**Comment by pliptor** [comment]
@Polyacene I didn't even know there was a `fit` function support. @Thanh-Binh How complex are the functions you are trying to fit?

**Comment by Thanh-Binh** [comment]
@pliptor it should be enough complex so that we can approximate/predict a human behavior, e.g how to accelerate or brake at some driving situations

**Comment by Polyacene** [comment]
@Thanh-Binh I've posted one plot before, and I played a little bit. I still havn't performed a systematic study.
Unsurprisingly it looks like a large number of samples and a large number of epochs is needed for
good approximations.
@pliptor One I idea I tried was adding approximations to the function as additional input. But the improvement was not too exciting. I still think this should work somehow.

**Comment by Polyacene** [comment]
@Thanh-Binh  Note, I just realized that my example code doesn't compile with the latest tiny-dnn version and I can't get it compile. The documented new  way

'''C++

  	using namespace tiny_dnn::layers;
	
	network<sequential> nn;
  	nn 	<< fully_connected_layer(1, NumHidden) << tanh()
 		<< fully_connected_layer(NumHidden, NumHidden)<< tanh()
 		<< fully_connected_layer(NumHidden, 1)<< tanh();
'''

seems not to work, and the identity activation (linear) seems to be gone.

**Comment by beru** [comment]
How about removing last `tanh`?

**Comment by Polyacene** [comment]
@beru It's actually already the first tanh() that makes troubles, as the compiler complains about a conflict with tanh(double).

It compiles with
'''C++

	network<sequential> nn;
  	nn 	<< fully_connected_layer(1, NumHidden) <<  tiny_dnn::activation::tanh()
 		<< fully_connected_layer(NumHidden, NumHidden)<< tiny_dnn::activation::tanh()
 		<< fully_connected_layer(NumHidden, 1)  << tiny_dnn::activation::tanh();
'''

However, what confuses me is that

'''C++

	for (int i = 0; i < nn.depth(); i++) 
	{
		std::cout << "#layer :" << i << "\n";
		std::cout << "#\ttype  :" << nn[i]->layer_type() << "\n";
		std::cout << "#\tinput :" << nn[i]->in_data_size() << "(" << nn[i]->in_data_shape() << ")\n";
		std::cout << "#\toutput:" << nn[i]->out_data_size() << "(" << nn[i]->out_data_shape() << ")\n" << std::endl;
	}

'''


now reports about 6 layers. So I'm not sure what kind of network is actually constructed.

**Comment by pliptor** [comment]
@Polyacene I haven't played with it since the API change. Maybe they have changed the way layers are counted. Displaying the network with [this](https://github.com/tiny-dnn/tiny-dnn/blob/master/docs/how_tos/How-Tos.md#visualize-graph-networks) might be helpful.

**Comment by beru** [comment]
@Polyacene I haven't tested but would this be equivalent with old network?

```cpp
network<sequential> nn;
nn 	<< fully_connected_layer(1, NumHidden) <<  tiny_dnn::activation::tanh()
	<< fully_connected_layer(NumHidden, NumHidden)<< tiny_dnn::activation::tanh()
	<< fully_connected_layer(NumHidden, 1)
;
```

**Comment by Polyacene** [comment]
@beru Then it looks like I get 5 layers. I think I'll prepare a new thread as I'm really confused.

**Comment by Polyacene** [comment]
@Beru, @Thanh-Binh Here's a version that compiles and works with the new layout of tiny-dnn.
There's no activation in the input as the convergence was slower. 
As clarified by @karandesai-96 the activation does count as a layer now, so everything is fine, except
that one should also specify the input size. Have fun.
[interpolate.C.txt](https://github.com/tiny-dnn/tiny-dnn/files/963704/interpolate.C.txt)

**Comment by Thanh-Binh** [comment]
@polyacene congratulation, i will test it today!

**Comment by karandesai-96** [comment]
Also, about the input shape of activation layers, this might help: https://github.com/tiny-dnn/tiny-dnn/pull/639

We are planning to extend for all layers.

**Comment by Polyacene** [comment]
@karandesai-96 In the following
```
       network<sequential> nn;
  	nn 	<< tiny_dnn::activation::tanh() << fully_connected_layer(1, 2 );
  	
  	nn 	<< fully_connected_layer(2, NumHidden )
		<< tiny_dnn::activation::tanh() << fully_connected_layer(NumHidden, NumHidden)
		<< tiny_dnn::activation::tanh(NumHidden) << fully_connected_layer(NumHidden, 1);
```

the second tanh() works fine, while the first tanh() leads to a crash,
which makes absolute sense as the dimension of it not specified.
It would have to use the dimension of the following layer.
It runs fine if the first `tanh()` is replaced by `tanh(1)`

**Comment by Polyacene** [comment]
@karandesai-96  I tried more fun stuff, like using 2 different activation functions in an activation layer.

```C++
	layers::concat    concat(2, 1);
	network<sequential> nn;

	nn 	<< fully_connected_layer(1, 2 );
	nn 	<< concat << ( tiny_dnn::activation::tanh(1), tiny_dnn::activation::relu(1) );
  	
  	nn 	<< fully_connected_layer(2, 1 );
```

I might have misunderstood the concept of `concat`, however the error message is interesting:

> terminate called after throwing an instance of 'tiny_dnn::nn_error'
  what():  layer dimension mismatch!
output size of Nth layer must be equal to input of (N+1)th layer
layerN:   fully-connected in:1([[1x1x1],[1x2x1],[2x1x1]]), out:2([[2x1x1]])
layerN+1:       concat in:2([[1x1x1],[1x1x1]]), out:2([[1x1x2]])
2 != 2
Aborted (core dumped)

To the best of my knowledge even in todays world we still have 2 == 2 .

**Comment by karandesai-96** [comment]
I know this 2 != 2 reason, I had it in my mind to fix it, but forgot later. Thanks for reminding.

You probably misinterpreted concat layer. It concatenates inputs. The network architecture you write, seems like concat layer should have been a `slice` layer instead.
